x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/InternalGeoLine.java

```diff
@@ -0,0 +1,200 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.spatial.search.aggregations;
+
+import org.apache.lucene.geo.GeoEncodingUtils;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.search.aggregations.InternalAggregation;
+import org.elasticsearch.search.sort.SortOrder;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * A single line string representing a sorted sequence of geo-points
+ */
+public class InternalGeoLine extends InternalAggregation {
+    private static final double SCALE = Math.pow(10, 6);
+
+    private long[] line;
+    private double[] sortVals;
+    private boolean complete;
+    private boolean includeSorts;
+    private SortOrder sortOrder;
+
+    InternalGeoLine(String name, long[] line, double[] sortVals, Map<String, Object> metadata, boolean complete,
+                    boolean includeSorts, SortOrder sortOrder) {
+        super(name, metadata);
+        this.line = line;
+        this.sortVals = sortVals;
+        this.complete = complete;
+        this.includeSorts = includeSorts;
+        this.sortOrder = sortOrder;
+    }
+
+    /**
+     * Read from a stream.
+     */
+    public InternalGeoLine(StreamInput in) throws IOException {
+        super(in);
+        this.line = in.readLongArray();
+        this.sortVals = in.readDoubleArray();
+        this.complete = in.readBoolean();
+        this.includeSorts = in.readBoolean();
+        this.sortOrder = SortOrder.readFromStream(in);
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeLongArray(line);
+        out.writeDoubleArray(sortVals);
+        out.writeBoolean(complete);
+        out.writeBoolean(includeSorts);
+        sortOrder.writeTo(out);
+    }
+
+    @Override
+    public InternalAggregation reduce(List<InternalAggregation> aggregations, ReduceContext reduceContext) {
+        int mergedSize = 0;
+        boolean complete = true;
+        boolean includeSorts = true;
+        for (InternalAggregation aggregation : aggregations) {
```

Comment:
As I understand, points on each aggregation are already sorted, I wonder if we should use a priority queue to retrieve the top points needed instead of copying all points into a new array?

Reply:
I initially thought this type of optimization was unnecessary, but now that you bring it up, If we are aggregating against a lot of shards, then the  reduce can blow up to be quiet a lot of additional memory here. Will update to do this!

Reply:
in my latest commit, I've switched to this strategy. I'm still debugging bugs in it. hope to resolve it soon!

Reply:
oh my. strategy worked. it was my test that had a bug. resolved!